From f32ca8c9a62e88df814d33d72797931e2f483ef6 Mon Sep 17 00:00:00 2001
From: Zhong Hongbo <hongbo.zhong@windriver.com>
Date: Tue, 11 Sep 2012 18:02:59 +0800
Subject: [PATCH 09/62] fsl_imx6: Add SDMA device init and
 sdma-imx6q-to1.bin.ihex

Source: Extract from vendor-drop package, L3.0.15_12.04.01_ER_source.tar.gz
(May,2012).

Signed-off-by: Zhong Hongbo <hongbo.zhong@windriver.com>
---
 arch/arm/mach-mx6/clock.c                 |    2 +-
 arch/arm/mach-mx6/mach-mx6q_sabrelite.c   |   27 ++++++++
 arch/arm/plat-mxc/include/mach/sdma.h     |    2 +
 firmware/Makefile                         |    3 +
 firmware/imx/sdma/sdma-imx6q-to1.bin.ihex |  100 +++++++++++++++++++++++++++++
 5 files changed, 133 insertions(+), 1 deletion(-)
 create mode 100644 firmware/imx/sdma/sdma-imx6q-to1.bin.ihex

diff --git a/arch/arm/mach-mx6/clock.c b/arch/arm/mach-mx6/clock.c
index 995a12d..81ed1d3 100644
--- a/arch/arm/mach-mx6/clock.c
+++ b/arch/arm/mach-mx6/clock.c
@@ -5032,7 +5032,7 @@ static struct clk_lookup lookups[] = {
 	_REGISTER_CLOCK(NULL, "ipg_clk", ipg_clk),
 	_REGISTER_CLOCK(NULL, "ipg_perclk", ipg_perclk),
 	_REGISTER_CLOCK(NULL, "spba", spba_clk),
-	_REGISTER_CLOCK("imx-sdma", NULL, sdma_clk[0]),
+	_REGISTER_CLOCK("imx35-sdma", NULL, sdma_clk[0]),
 	_REGISTER_CLOCK(NULL, "gpu2d_axi_clk", gpu2d_axi_clk),
 	_REGISTER_CLOCK(NULL, "gpu3d_axi_clk", gpu3d_axi_clk),
 	_REGISTER_CLOCK(NULL, "pcie_axi_clk", pcie_axi_clk),
diff --git a/arch/arm/mach-mx6/mach-mx6q_sabrelite.c b/arch/arm/mach-mx6/mach-mx6q_sabrelite.c
index 2817d2c..bb99f7d 100644
--- a/arch/arm/mach-mx6/mach-mx6q_sabrelite.c
+++ b/arch/arm/mach-mx6/mach-mx6q_sabrelite.c
@@ -284,11 +284,37 @@ static const struct fec_platform_data fec_data __initconst = {
 	.phy	= PHY_INTERFACE_MODE_RGMII,
 };
 
+static struct sdma_script_start_addrs imx6q_sdma_script = {
+	.ap_2_ap_addr = 642,
+	.uart_2_mcu_addr = 817,
+	.mcu_2_app_addr = 747,
+	.per_2_per_addr = 6331,
+	.uartsh_2_mcu_addr = 1032,
+	.mcu_2_shp_addr = 960,
+	.app_2_mcu_addr = 683,
+	.shp_2_mcu_addr = 891,
+	.spdif_2_mcu_addr = 1100,
+	.mcu_2_spdif_addr = 1134,
+	.mcu_2_ssish_addr = 6242,
+	.ssish_2_mcu_addr = 6678,
+};
+
+static struct sdma_platform_data imx6q_sdma_pdata __initdata = {
+	.fw_name = "imx/sdma/sdma-imx6q-to1.bin",
+	.script_addrs = &imx6q_sdma_script,
+};
+
 static void __init fixup_mxc_board(struct tag *tags, char **cmdline,
 				   struct meminfo *mi)
 {
 }
 
+static void __init mx6_soc_init(void)
+{
+	/* i.mx6q has the i.mx35 type sdma */
+	imx_add_imx_sdma("imx35-sdma", MX6Q_SDMA_BASE_ADDR, MX6Q_INT_SDMA, &imx6q_sdma_pdata);
+}
+
 /*!
  * Board specific initialization.
  */
@@ -299,6 +325,7 @@ static void __init mx6_sabrelite_board_init(void)
 
 	phy_register_fixup_for_uid(PHY_ID_KSZ9021, MICREL_PHY_ID_MASK,
 					   ksz9021rn_phy_fixup);
+	mx6_soc_init();
 
 	mx6q_sabrelite_init_uart();
 	imx6q_add_fec(&fec_data);
diff --git a/arch/arm/plat-mxc/include/mach/sdma.h b/arch/arm/plat-mxc/include/mach/sdma.h
index 3a39428..19cfa9a 100644
--- a/arch/arm/plat-mxc/include/mach/sdma.h
+++ b/arch/arm/plat-mxc/include/mach/sdma.h
@@ -43,6 +43,8 @@ struct sdma_script_start_addrs {
 	s32 dptc_dvfs_addr;
 	s32 utra_addr;
 	s32 ram_code_start_addr;
+	s32 mcu_2_ssish_addr;
+	s32 ssish_2_mcu_addr;
 };
 
 /**
diff --git a/firmware/Makefile b/firmware/Makefile
index 0d15a3d..f7db953 100644
--- a/firmware/Makefile
+++ b/firmware/Makefile
@@ -48,6 +48,9 @@ fw-shipped-$(CONFIG_CHELSIO_T3) += cxgb3/t3b_psram-1.1.0.bin \
 				   cxgb3/ael2005_opt_edc.bin \
 				   cxgb3/ael2005_twx_edc.bin \
 				   cxgb3/ael2020_twx_edc.bin
+ifdef CONFIG_IMX_SDMA
+fw-shipped-$(CONFIG_SOC_IMX6Q) += imx/sdma/sdma-imx6q-to1.bin
+endif
 fw-shipped-$(CONFIG_DRM_MGA) += matrox/g200_warp.fw matrox/g400_warp.fw
 fw-shipped-$(CONFIG_DRM_R128) += r128/r128_cce.bin
 fw-shipped-$(CONFIG_DRM_RADEON) += radeon/R100_cp.bin radeon/R200_cp.bin \
diff --git a/firmware/imx/sdma/sdma-imx6q-to1.bin.ihex b/firmware/imx/sdma/sdma-imx6q-to1.bin.ihex
new file mode 100644
index 0000000..b969928
--- /dev/null
+++ b/firmware/imx/sdma/sdma-imx6q-to1.bin.ihex
@@ -0,0 +1,100 @@
+:1000000053444D4101000000010000001C000000AD
+:1000100025000000B000000078050000820200000A
+:10002000FFFFFFFF00000000FFFFFFFFFFFFFFFFDC
+:10003000FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFD0
+:10004000FFFFFFFFFFFFFFFF6A1A0000FFFFFFFF38
+:10005000EB020000BB180000FFFFFFFF08040000D8
+:10006000FFFFFFFFC0030000FFFFFFFFFFFFFFFFD9
+:10007000FFFFFFFFAB020000FFFFFFFF7B0300005D
+:10008000FFFFFFFFFFFFFFFF4C0400006E040000B6
+:10009000FFFFFFFF00180000FFFFFFFFFFFFFFFF54
+:1000A000000000000018000062180000161A00008E
+:1000B000E3C1DB57E35FE357F352016A8F00D500DA
+:1000C000017D8D00A005EB5D7804037D79042C7D16
+:1000D000367C79041F7CEE56000F6006057D0965AD
+:1000E000437E0A62417E20980A623E7E09653C7E1C
+:1000F00012051205AD026007037DFB55D36D2B98E9
+:10010000FB55041DD36DC86A2F7F011F03200048D3
+:10011000E47C5398FB55D76D150005780962C86AD1
+:100120000962C86AD76D5298FB55D76D1500150046
+:1001300005780A62C86A0A62C86AD76D5298FB5588
+:10014000D76D15001500150005780B62C86A0B62A3
+:10015000C86AD76D097CDF6D077F0000EB55004D45
+:10016000077DFAC1E35706980700CC680C6813C2F4
+:100170000AC20398D9C1E3C1DB57E35FE357F352E7
+:10018000216A8F00D500017D8D00A005EB5DFB5637
+:100190007804037D79042A7D317C7904207C700BFE
+:1001A0001103EB53000F6003057D0965377E0A627A
+:1001B000357E86980A62327E0965307E1205120508
+:1001C000AD026007027C065A8E98265A277F011FCF
+:1001D00003200048E87C700B11031353AF981500FF
+:1001E00004780962065A0962265AAE98150015006D
+:1001F00004780A62065A0A62265AAE98150015005B
+:10020000150004780B62065A0B62265A077C000020
+:10021000EB55004D067DFAC1E357699807000C685D
+:1002200013C20AC26698700B110313536C07017C4A
+:10023000D9C1FB5E8A066B07017CD9C1F35EDB592D
+:10024000D3588F0110010F398B003CC12B7DC05A50
+:10025000C85B4EC1277C88038906E35CFF0D11054E
+:10026000FF1DBC053E07004D187D700811007E077C
+:10027000097D7D07027D2852E698F852DB54BC02C6
+:10028000CC02097C7C07027D2852EF98F852D354A7
+:10029000BC02CC02097D0004DD988B00C052C8531B
+:1002A00059C1D67D0002CD98FF08BF007F07157D9C
+:1002B0008804D500017D8D00A005EB5D8F02120240
+:1002C0001202FF3ADA05027C3E071899A402DD0209
+:1002D000027D3E0718995E071899EB559805EB5D6E
+:1002E000F352FB546A07267D6C07017D55996B0715
+:1002F000577C6907047D6807027D010E2F9993588A
+:10030000D600017D8E009355A005935DA00602786E
+:100310000255045D1D7C004E087C6907037D025573
+:10032000177E3C99045D147F890693500048017D37
+:100330002799A099150006780255045D4F070255CC
+:10034000245D2F07017CA09917006F07017C012015
+:1003500093559D000700A7D9F598D36C6907047DD4
+:100360006807027D010E64999358D600017D8E00C6
+:100370009355A005935DA00602780255C86D0F7CC9
+:10038000004E087C6907037D0255097E7199C86D8E
+:10039000067F890693500048017D5C99A0999A993F
+:1003A000C36A6907047D6807027D010E8799935827
+:1003B000D600017D8E009355A005935DA0060278BE
+:1003C000C865045D0F7C004E087C6907037DC86525
+:1003D000097E9499045D067F890693500048017D4B
+:1003E0007F99A09993559D000700FF6CA7D9F598B8
+:1003F0000000E354EB55004D017CF598DD98E35483
+:10040000EB55FF0A1102FF1A7F07027CA005B49981
+:100410009D008C05BA05A0051002BA04AD04540471
+:100420000600E3C1DB57FB52C36AF352056A8F0033
+:10043000D500017D8D00A005EB5D7804037D790476
+:100440002B7D1E7C7904337CEE56000FFB55600734
+:10045000027DC36DD599041DC36DC8623B7E6006E5
+:10046000027D10021202096A357F1202096A327F88
+:100470001202096A2F7F011F03200048E77C099AB6
+:10048000FB55C76D1500150015000578C8620B6A8D
+:10049000C8620B6AC76D089AFB55C76D1500150039
+:1004A0000578C8620A6AC8620A6AC76D089AFB556D
+:1004B000C76D15000578C862096AC862096AC76D08
+:1004C000097C286A077F0000EB55004D057DFAC1C5
+:1004D000DB57BF9977C254040AC2BA99D9C1E3C1A4
+:1004E000DB57F352056A8F00D500017D8D00A00512
+:1004F000FB567804037D7904297D1F7C79042E7CCA
+:10050000E35D700D1105ED55000F6007027D065289
+:10051000329A2652337E6005027D10021202096A69
+:100520002D7F1202096A2A7F1202096A277F011FA2
+:1005300003200048EA7CE3555D9A1500150015007C
+:10054000047806520B6A26520B6A5C9A1500150055
+:10055000047806520A6A26520A6A5C9A15000478E0
+:100560000652096A2652096A097C286A077F000038
+:10057000DB57004D057DFAC1DB571B9A77C2540447
+:100580000AC2189AE3C1DB57F352056AFB568E0282
+:10059000941AC36AC8626902247D941EC36ED36E26
+:1005A000C8624802C86A9426981EC36ED36EC86299
+:1005B0004C02C86A9826C36E981EC36EC8629826FD
+:1005C000C36E6002097CC8626E02247D096A1E7FC8
+:1005D0000125004D257D849A286A187F04627AC21D
+:1005E000B89AE36E8F00D805017D8D00A005C86222
+:1005F0006E02107D096A0A7F0120F97C286A067F55
+:100600000000004D0D7DFAC1DB576E9A07000462B1
+:100610000C6AB59A286AFA7F04627AC258045404B4
+:08062000286AF47F0AC26B9AFC
+:00000001FF
-- 
1.7.9.7

